### PR TITLE
Encrypted zksnark prototype

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
## Changelog

* Created encrypted Zksnark prototype that encrypts powers
* Split from unencrypted zksnark example
* Exposed useful transcript trait

## Motivation

In continuing on with the zksnark mockup, we wanted to add encryption of powers using the eliptic curve group

## Future Work
* Create necessary shifts to ensure powers & resistance against brute force 
* Refactor primitives like the polynomials to be able to be used centrally and to note use i64 directly
* Secure calculations

### CODE REVIEW

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+pub struct Root {
+    // a in ax+b
+    a: Scalar,
+    // b in ax+b
+    b: Scalar,
+}
+
+impl Root {
+    /// Create a new root from i64 values
+    pub fn new(a: i64, b: i64) -> Result<Self, Error> {
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:18]: Optimally we should change this to something else? Like just initialize with scalars or have any integers be able to translate into scalars/create "from" implementations. For the tutorial maybe this okay, but I'd like to see a follow-on PR that changes this.

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+/// Object providing an encrypted list of scalars encrypted by multiplying them by an
+/// elliptic curve (making them discrete log hard to calculate) and an evaluation of
+/// the public roots of the target polynomial that is able to verify prover responses.
+pub struct EncryptedChallenge {
+    /// List of Ristretto curve points created by multiplying the secret scalar by the
+    /// Ristretto basepoint
+    pub encrypted_powers: Vec<RistrettoPoint>,
+    // Public roots of the target polynomial evaluate at the secret scalar
+    ts: Scalar,
+}
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:51]: Potentially we should have a scalar generation function attached this?

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        Self {
+            encrypted_powers,
+            ts,
+        }
+    }
+
+    /// Verify a prover response by multiplying t(s) the provers calculated curve points
+    pub fn verify_response(&self, response: &EncryptedResponse) -> bool {
+        self.ts * response.hx == response.px
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:72]: This is better

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            }
+            //println!("coefficients: {:?}", coefficients);
+        }
+        coefficients.reverse();
+        Ok(Self {
+            roots,
+            coefficients,
+            num_public_roots,
+        })
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:138]: This process could be made more efficient, please address it in a follow-on PR

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+    /// Evaluate polynomial at given encrypted powers
+    pub fn eval(&self, powers: &EncryptedChallenge) -> RistrettoPoint {
+        powers
+            .encrypted_powers
+            .iter()
+            .zip(self.coefficients.iter())
+            .map(|(p, c)| p * c)
+            .sum()
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:161]: This should probably be made non-public, and tests of correct polynomial evaluation should be made another way.

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+    /// Get degree of polynomial
+    pub fn degree(&self) -> usize {
+        self.roots.len()
+    }
+
+    /// calculate encrypted polynomial
+    pub fn generate_response(&self, powers: &EncryptedChallenge) -> EncryptedResponse {
+        let px = self.eval(powers);
+        let h = Polynomial::new(self.roots[self.num_public_roots..].to_vec(), 1).unwrap();
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:148]: This unwrap here isn't the best thing. Perhaps these polynomial coefficients should be baked into the object instead.

File: applied-crypto-examples/src/encrypted_zksnark.rs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+            Root::new(2, 4).unwrap(),
+        ];
+
+        let polynomial = Polynomial::new(roots, 2).unwrap();
+        let scalar = Scalar::from(5u64);
+        let challenge = EncryptedChallenge::new(&scalar, &polynomial);
+
+        assert_eq!(polynomial.degree(), 3);
+        assert_eq!(polynomial.eval(&challenge), Scalar::from(2058u64) * G);
+    }
```
Comment (GlitchyBritches) [inline applied-crypto-examples/src/encrypted_zksnark.rs:217]: This should be done without using a public eval.

**Review by GlitchyBritches — APPROVED** [review]
Feedback should be addressed in a PR-follow-on

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 1 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:18]: Yeah, probably just scalars and have a from function that translates between i64. If this turns into a productoin version, we might make it more robust.

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 2 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:51]: Possibly yes, especially if we're looking to create non-interactivity.

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 4 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:138]: Indeed, will address in the follow-on

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 5 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:161]: Agreed!

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 6 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:148]: Agreed on that point

File: applied-crypto-examples/src/encrypted_zksnark.rs (hunk 7 above)
Comment (iamalwaysuncomfortable) [reply applied-crypto-examples/src/encrypted_zksnark.rs:217]: Agreed!

**Pull request merged**